File: src/etch/bindings/java/util/MapSerializer.java
/* $Id$
 *
 * Created by sccomer on Apr 14, 2008.
 *
 * Copyright (c) 2005 Metreos, Inc. All rights reserved.
 */

package etch.bindings.java.util;

import java.util.HashMap;
import java.util.Map;

import etch.bindings.java.msg.Field;
import etch.bindings.java.msg.ImportExportHelper;
import etch.bindings.java.msg.StructValue;
import etch.bindings.java.msg.Type;
import etch.bindings.java.support.Class2TypeMap;
import etch.bindings.java.support.Validator_object;

/**
 * Serializer for generic Map<Object, Object>.
 */
public class MapSerializer implements ImportExportHelper
{
	private final static String FIELD_NAME = "keysAndValues";

	/**
	 * Defines custom fields in the value factory so that the importer
	 * can find them.
	 * @param type
	 * @param class2type 
	 */
	public static void init( Type type, Class2TypeMap class2type )
    {
		Field field = type.getField( FIELD_NAME );
		
		// quick lookup to serialize this class.
		class2type.put( HashMap.class, type );
		
		// type for arrays.
		type.setComponentType( Map.class );
		
		type.setImportExportHelper( new MapSerializer( type, field ) );
		type.putValidator( field, Validator_object.get( 1 ) );
		type.lock();
    }

	/**
	 * @param type
	 * @param field
	 */
	public MapSerializer( Type type, Field field )
	{
		this.type = type;
		this.field = field;
	}
	
	private final Type type;
	
	private final Field field;
	
	public Object importValue( StructValue struct )
	{
		struct.checkType( type );
		Object[] keysAndValues = (Object[]) struct.get( field );
		int n = keysAndValues.length;
		
		Map<Object, Object> map = new HashMap<Object, Object>( (n*2+2)/3, 0.75f ); // (n/2)*4/3
		int index = 0;
		while (index < n)
		{
			Object key = keysAndValues[index++];
			Object value = keysAndValues[index++];
			map.put( key, value );
		}
		
		return map;
	}

	@SuppressWarnings("unchecked")
	public StructValue exportValue( Object value )
	{
		Map<Object, Object> map = (Map<Object, Object>) value;
		
		Object[] keysAndValues = new Object[map.size()*2];
		int index = 0;
		for (java.util.Map.Entry<Object, Object> me: map.entrySet())
		{
			keysAndValues[index++] = me.getKey();
			keysAndValues[index++] = me.getValue();
		}
		
		StructValue struct = new StructValue( type );
		struct.put( field, keysAndValues );
		return struct;
	}
}

File: src/etch/bindings/csharp/util/MapSerializer.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;
using Etch.Msg;
using Etch.Support;

namespace Etch.Util
{
    public class MapSerializer : ImportExportHelper
    {
        private static String FIELD_NAME = "keysAndValues";

         /// <summary>
        /// Defines custom fields in the value factory so that the importer
        /// can find them.
        /// </summary>
        /// <param name="type"></param>
        /// <param name="class2type"></param>
        public static void Init( XType type, Class2TypeMap class2type )
        {
            Field field = type.GetField( FIELD_NAME );
      /*      class2type.Add( typeof( Dictionary<object,object> ), type );
            type.SetClass(typeof(Dictionary<object, object>)); */
            class2type.Add(typeof(Hashtable), type);
            type.SetComponentType(typeof(Hashtable));
            type.SetImportExportHelper( new MapSerializer( type, field ) );
            type.PutValidator( field, Validator_object.Get( 1 ) );
            type.Lock();
        }

        public MapSerializer( XType type, Field field )
        {
            this.type = type;
            this.field = field;
        }

        private readonly XType type;
        private readonly Field field;


        public override Object ImportValue( StructValue sv )
        {
     //      Dictionary<object,object> map = new Dictionary<object,object>();
           Hashtable map = new Hashtable();
            Object[] keysAndValues = ( Object[] ) sv.Get( field );
            int n = keysAndValues.Length;
            int index = 0;
            while ( index < n )
            {
                object key = ( object ) keysAndValues[ index++ ];
                object value = ( object ) keysAndValues[ index++ ];
                map.Add( key, value );
            }

            return map;
        }

        public override StructValue ExportValue( Object value )
        {
       //     Dictionary<object, object> map = (Dictionary<object, object>)value;

            Hashtable map = (Hashtable)value;

            Object[] keysAndValues = new Object[ map.Count * 2];
            int index = 0;

            foreach ( DictionaryEntry me in map )
            {
                keysAndValues[ index++ ] = me.Key;
                keysAndValues[ index++ ] = me.Value;
            }

            StructValue sv = new StructValue( type );
            sv.Add( field, keysAndValues );
            return sv;

        }

    }
}
